openrtx/src/ui/default/ui_menu.c

```diff
@@ -731,6 +761,168 @@ void _ui_drawMenuGPS()
+                    gfx_drawRect((point_t){x_start - bar_width, y_pos}, bar_width, bar_height, bar_color, true);
+                }
+            }
+        }
+        state.spectrum_shouldRefresh = false;
+    }
+    // Print small text on the left of the spectrum with the start frequency
+    sniprintf(freq_str, sizeof(freq_str), "%lu.%02lu",
+            (unsigned long)(state.spectrum_startFreq / 100000),
+            (unsigned long)(state.spectrum_startFreq % 100000 / 1000));
+    gfx_print((point_t){4, 8}, FONT_SIZE_5PT, TEXT_ALIGN_LEFT,
+                color_white, freq_str);
+    // Print small text on the right of the spectrum with the end frequency
+    sniprintf(freq_str, sizeof(freq_str), "%lu.%02lu",
+            (unsigned long)((state.spectrum_startFreq + NUMBER_BARS * spectrumStep) / 100000),
+            (unsigned long)((state.spectrum_startFreq + NUMBER_BARS * spectrumStep) % 100000 / 1000));
+    gfx_print((point_t){layout.horizontal_pad, 126}, FONT_SIZE_5PT, TEXT_ALIGN_LEFT,
+                color_white, freq_str);
+
+    if(state.spectrum_currentPart == 1){
+        gfx_render();
+        // Shift waterfall lines up by copying pixels from the framebuffer
+        // The waterfall area is approximately from y=117 to y=160 (43 lines)
+        const int waterfall_start_y = 0;
+        const int waterfall_end_y = 160;
+        
+        // Shift all lines up by copying each pixel from the line below
+        for(int y = waterfall_start_y; y < waterfall_end_y - 1; y++)
+        {
+            for(int x = 0; x < 128 && x < CONFIG_SCREEN_WIDTH; x++)
+            {
+                // Copy pixel from line below (y+1) to current line (y)
+                point_t src_pos = {x, y + 1};
+                point_t dst_pos = {x, y};
+                // For now, just draw white pixels for testing - will fix waterfall later
+                gfx_setPixel(dst_pos, (color_t){255,255,255,255});
+            }
+        }
+        
+        // Draw new waterfall line at the bottom of the waterfall area
+        int new_line_y = waterfall_end_y - 1;
+        for(int i = 0; i < NUMBER_BARS && i < 64; i++)
+        {
+            uint8_t height = state.spectrum_data[NUMBER_BARS-(i+1)];
+            bar_color = getColorFromLevel(height * (state.settings.spectrum_multiplier-1) + height);
+            
+            // Draw two pixels wide for each spectrum bar (like the original code)
+            point_t pos1 = {i * 2, new_line_y};
+            point_t pos2 = {i * 2 + 1, new_line_y};
+            
+            if(pos1.x < CONFIG_SCREEN_WIDTH && pos1.x < 128)
+                gfx_setPixel(pos1, bar_color);
+            if(pos2.x < CONFIG_SCREEN_WIDTH && pos2.x < 128)
+                gfx_setPixel(pos2, bar_color);
+        }
+        
+        state.spectrum_currentWFLine = ((state.spectrum_currentWFLine+1) % (waterfall_end_y - waterfall_start_y));
+    }
+    state.spectrum_currentPart = (state.spectrum_currentPart + 1) % 2;
+    ui_updateFSM(NULL);
```

Comment:
Remove this function call, `ui_updateFSM` is already called by the main UI thread. Moreover, passing NULL as parameter may crash the firmware